Packager docs (#4303)
quiltdata/quilt: docs/Catalog/Packaging.md

> <!--pytest-codeblocks:skipfile-->

Reviewer: maybe skip specific code block?

Author: @greptilai
Remove this skipfile, and instead just skip the actual pytest-codeblocks below

Author: ```suggestion
```

in Amazon S3. It serves as a key component in Quilt's SDMS (Scientific Data
Management System) strategy, enabling automated data ingestion and
standardization. It consists of:

1. Admin Settings GUI to enable package creation based on notifications from:
   1. AWS Health Omics
   2. Nextflow workflows using  `nf-prov`'s WRROC ([Workflow Run
      RO-Crate](https://www.researchobject.org/workflow-run-crate/)) format
2. An SQS queue that will process package descriptions
3. Documentation for creating custom EventBridge rules to invoke that queue

## Admin Settings

The simplest way to enable package creation is through the Admin Settings GUI,
which supports the following built-in event sources:

![Admin Settings](../imgs/package-admin-gui.png)

### AWS Health Omics

When enabled, this will create a package from the `runOutputUri` provided in a
`aws.omics` completion event. For example, if the `runOutputUri` is
`s3://quilt-example/omics-quilt/3395667`, the package will be created in that
same bucket with the name `omics-quilt/3395667`.

### Workflow Run RO-Crate

When enabled, this will create a package when ingesting any folder containing an

> `ro-crate-manifest.json`.  Ingest happens when the bucket is added to the stack,

Reviewer: > Ingest happens when the bucket is added to the stack,

no

Author: ```suggestion
When enabled, this will create a package when indexing any folder containing an
`ro-crate-manifest.json`.  Indexing happens when the bucket is added to the stack,
```

Is indexing a better term? Do we not get these notifications when a bucket is added?

Reviewer: when we add a bucket we generate "S3 events" only for indexer queue not for our bus

profiles, which are supported in the latest versions of
[nf-prov](https://github.com/nextflow-io/nf-prov). You will need to explicitly
configure `nf-prov` to use wrroc, by using a `nextflow.config` file [like
this](https://github.com/famosab/wrrocmetatest):

```groovy
plugins {
 id 'nf-prov@1.4.0'
}

prov {
 enabled = true
 formats {
   wrroc {
     file = "${params.outdir}/ro-crate-metadata.json"
     overwrite = true
     agent {
       name = "John Doe"
       orcid = "https://orcid.org/0000-0000-0000-0000"
     }
      license = "https://spdx.org/licenses/MIT"
      profile = "provenance_run_crate"
   }
 }
}
```

Note that Research Objects identify people using an ORCID iD, which anyone can
get for free at [the ORDiD website](https://orcid.org/).

The package will be created in the same bucket as the `outdir`, with the package
name inferred from the folder key. For example, if the key is `my/s3/folder`,
the package name will be `my_s3/folder`.

## SQS Message Processing

The primary interface to the Packaging Engine is through the
SQS queue in the same account and region as your Quilt stack,
listed in `PackagerQueue` under the Outputs. The queue URL will look something like:

> ```url

Reviewer: I think markdownlint [recommends](https://github.com/DavidAnson/markdownlint/blob/main/doc/md040.md) this
```suggestion
```text
```

https://sqs.us-east-1.amazonaws.com/XXXXXXXXXXXX/PackagerQueue-XXXXXXXXXXXX
```

The body of the message is the stringified JSON of a package description.
There is only one required parameter:

```json
{
  "source_prefix": "s3://data_bucket/source/folder/metadata.json"
}
```

This will create a package in the same bucket as the source folder, with the
package name being inferred from the source URI.

Optionally, you can control the package name, metadata, and other settings by
explicitly specifying any of the following fields:

```jsonc
{

>   "source_prefix": "s3://data_bucket/source/folder/",

Reviewer: it should be noted that we remove the last path segment if it doesn't end with slash

Author: @greptileai
clarify that if the source_prefix URI ends in a slash, it will be treated as a folder. Otherwise, we will treat the last path segment as a file and package the parent

Author: Fixed manually

  "registry": "package_bucket",  // may be the same as `data_bucket`
  "package_name": "prefix/suffix",

>   "metadata": { "key": "value" },  // dictionary

Reviewer: `object` seems the right term

Author: ```suggestion
  "metadata": { "key": "value" },  // object
```

  "metadata_uri": "metadata.json", // alternative to `metadata`, relative or absolute
  "commit_message": "Commit message for the package revision", // string
  "workflow": "alpha", // name of a valid metadata workflow
}
```

### SendMessage API

If you have appropriate IAM permissions, and the SQS URL, you can send a message
to the queue using the AWS SDK or the AWS CLI. Here is an example using the AWS
CLI:

> 

Author: ```suggestion

<!--pytest.mark.skip-->
```

```bash
export QUEUE_URL=https://sqs.us-east-1.amazonaws.com/XXXXXXXXXXXX/PackagerQueue-XXXXXXXXXXXX
aws sqs send-message --queue-url $QUEUE_URL \
--message-body '{"source_prefix":"s3://data_bucket/source/folder/"}'
```

## Custom EventBridge Rules

EventBridge event in your account (from any bus, in any region) into a

> conforming SQS message.

Reviewer: is this complete sentence? 🤔

Author: @greptileai
Complete the sentence

### Example: Event-Driven Packaging (EDP)

> EDP is a high-end add-on to Quilt that coalesces multiple S3 uploads into a

Reviewer: reference to EDP docs?

Author: EDP linnk

single `package-objects-ready` event, where it infers the appropriate top-level
folder. When ready, it creates an event like this on its own EventBridge bus:

```json
{
  "version":"0",
  "id":"XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
  "detail-type":"package-objects-ready",
  "source":"com.quiltdata.edp",
  "account":"XXXXXXXXXXXX",
  "time":"2022-12-08T20:01:34Z",
  "region":"us-east-1",
  "resources":[
    "arn:aws:s3:::bucket-name"
  ],
  "detail":{
    "version":"0.1",
    "bucket":"bucket-name",
    "prefix":"prefix-path-1/prefix-path-2/"
  }
}
```

Here is an example of an EventBridge rule that will use the above event
to trigger packager queue:

> ```json

Reviewer: is this part of CloudFormation template or?

Author: Not sure I understand the question.  This is just an example for the customers to use/adapt; we are not creating this as part of the stack.

Reviewer: where do I put this JSON?

Author: Ah, my bad for believing ChatGPT.  I will split into the pattern and target.

Author: Actually, changed to boto3 code.

describing research data.  The Workflow Run working group adds three additional
profiles, which are supported in the latest versions of
[nf-prov](https://github.com/nextflow-io/nf-prov). You will need to explicitly
configure `nf-prov` to use wrroc, by using a `nextflow.config` file [like
this](https://github.com/famosab/wrrocmetatest):

```groovy
plugins {
 id 'nf-prov@1.4.0'
}

prov {
 enabled = true
 formats {
   wrroc {
     file = "${params.outdir}/ro-crate-metadata.json"
     overwrite = true
     agent {
       name = "John Doe"
       orcid = "https://orcid.org/0000-0000-0000-0000"
     }
      license = "https://spdx.org/licenses/MIT"
      profile = "provenance_run_crate"
   }
 }
}
```

Note that Research Objects identify people using an ORCID iD, which anyone can
get for free at [the ORDiD website](https://orcid.org/).

The package will be created in the same bucket as the `outdir`, with the package
name inferred from the folder key. For example, if the key is `my/s3/folder`,
the package name will be `my_s3/folder`.

## SQS Message Processing

The primary interface to the Packaging Engine is through the
SQS queue in the same account and region as your Quilt stack,

> listed in `PackagerQueue` under the Outputs. The queue URL will look something like:

Author: @sir-sigurd Were you planning to add both Url and Arn for the SQS queue?

Reviewer: I was and did 😄

```text
https://sqs.us-east-1.amazonaws.com/XXXXXXXXXXXX/PackagerQueue-XXXXXXXXXXXX
```

The body of the message is the stringified JSON of a package description.
There is only one required parameter:

```json
{
  "source_prefix": "s3://data_bucket/source/folder/metadata.json"
}
```

This will create a package in the same bucket as the source folder, with the
package name being inferred from the source URI.

Optionally, you can control the package name, metadata, and other settings by
explicitly specifying any of the following fields:

```jsonc
{
  "source_prefix": "s3://data_bucket/source/folder/",  // If URI ends with slash, treats entire path as folder. Without slash, treats last segment as file and packages parent folder
  "registry": "package_bucket",  // may be the same as `data_bucket`
  "package_name": "prefix/suffix",
  "metadata": { "key": "value" },  // object

>   "metadata_uri": "metadata.json", // alternative to `metadata`, relative or absolute

Author: @sir-sigurd What happens if they specify both types of metadata? Do we merge, or pick one?

Reviewer: fail

Author: > The job will fail if you try to specify both `metadata` and `metadata_uri`.
